model/AdyenNotification.php

```diff
@@ -0,0 +1,216 @@
+            . ' AND `entity_id` = "' . (int)$id . '"';
+
+        return $this->dbInstance->update(self::$tableName, $data, $where);
+    }
+
+    /**
+     * @param $notification
+     * @throws \Exception
+     */
+    public function insertNotification($notification)
+    {
+        $data = array();
+        if (isset($notification['pspReference'])) {
+            $data['pspreference'] = pSQL($notification['pspReference']);
+        }
+        if (isset($notification['originalReference'])) {
+            $data['original_reference'] = pSQL($notification['originalReference']);
+        }
+        if (isset($notification['merchantReference'])) {
+            $data['merchant_reference'] = pSQL($notification['merchantReference']);
+        }
+        if (isset($notification['eventCode'])) {
+            $data['event_code'] = pSQL($notification['eventCode']);
+        }
+        if (isset($notification['success'])) {
+            $data['success'] = pSQL($notification['success']);
+        }
+        if (isset($notification['paymentMethod'])) {
+            $data['payment_method'] = pSQL($notification['paymentMethod']);
+        }
+        if (isset($notification['amount'])) {
+            $data['amount_value'] = pSQL($notification['amount']['value']);
+            $data['amount_currency'] = pSQL($notification['amount']['currency']);
+        }
+        if (isset($notification['reason'])) {
+            $data['reason'] = pSQL($notification['reason']);
+        }
+
+        if (isset($notification['additionalData'])) {
+            $data['additional_data'] = pSQL(serialize($notification['additionalData']));
+        }
+        if (isset($notification['done'])) {
+            $data['done'] = pSQL($notification['done']);
+        }
+
+        // do this to set both fields in the correct timezone
+        $date = new \DateTime();
+        $data['created_at'] = $date->format('Y-m-d H:i:s');
+        $data['updated_at'] = $date->format('Y-m-d H:i:s');
+
+        $this->dbInstance->insert(self::$tableName, $data);
+    }
+
+    /**
+     * If notification is already saved ignore it
+     *
+     * @param $notification
+     * @return mixed
+     */
+    public function isDuplicate($notification)
+    {
+        $pspReference = trim($notification['pspReference']);
+        $eventCode = trim($notification['eventCode']);
+        $success = trim($notification['success']);
+
+        $sql = 'SELECT * FROM ' . _DB_PREFIX_ . self::$tableName
+            . ' WHERE `pspreference` = "' . pSQL($pspReference) . '"'
+            . ' AND `event_code` = "' . pSQL($eventCode) . '"'
+            . ' AND `success` = "' . pSQL($success) . '"';
+
+        $originalReference = null;
+        if (!empty($notification['originalReference'])) {
+            $originalReference = trim($notification['originalReference']);
+            $sql .= ' AND `original_reference` = "' . pSQL($originalReference) . '"';
+        }
+
+        return $this->dbInstance->getValue($sql);
+    }
+}
```

Comment:
I really love this entire file. Thanks for putting the effort in this kind of organization.

Reply:
Thanks for noticing ;)